### PR TITLE
feat: device discovery + bulk config

### PR DESCRIPTION
**Pull request by alex-sandrk**

### CODE REVIEW

File: src/plugins/devices/cgminer-api.js [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+  isSocketOpen() {
+    return this.socket.readyState === 'open'
+  }
+
+  /**
+   * Performs connection over the socket with initial data
+   * @returns {Promise<void>}
+   */
+  async reconnect() {
+    this.socket = new Socket()
```
Comment (lsheva) [inline src/plugins/devices/cgminer-api.js:39]: allowHalfOpen ?

File: src/plugins/devices/configuration-strategies/tcp-strategy.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+   * @param {String} poolUser
+   * @returns {Promise<boolean>} Returns true if successfully updated configuration
+   */
+  async setPool(pool, poolUser) {
+    try {
+      const api = new CGMinerApi()
+      await api.connect({ host: this.host, abort: this.abort })
+
+      const result = await api.addPool(pool, poolUser)
+      const status = result.STATUS[0].STATUS
```
Comment (lsheva) [inline src/plugins/devices/configuration-strategies/tcp-strategy.js:38]: if result is undefined this is going to fail

please consider throwing error inside `api.addPool`

File: src/plugins/devices/miner-configurator.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+const setPool = async (host, poolUrl, abort, onUpdate) => {
+  const strategy = await ConfigurationStrategyFactory.createStrategy(
+    host,
+    abort
+  )
+  if (!strategy) {
+    throw new Error('No available configuration strategy')
+  }
+  const poolUser = `proxy.${host.split('.').slice(-2).join('.')}`
+  const result = await strategy.setPool(poolUrl, poolUser)
```
Comment (lsheva) [inline src/plugins/devices/miner-configurator.js:22]: if setPool fails there will be no error message on client

File: src/plugins/devices/configuration-strategies/tcp-strategy.js (hunk 2 above)
Comment (lsheva) [reply src/plugins/devices/configuration-strategies/tcp-strategy.js:38]: i would even consider moving the parsing logic inside api

therefore the method would return Promise<void> or rejection

and that is I think a preferable way to return results from success/error functions

Comment (alex-sandrk) [reply src/plugins/devices/configuration-strategies/tcp-strategy.js:38]: updated

File: src/plugins/devices/miner-configurator.js (hunk 3 above)
Comment (alex-sandrk) [reply src/plugins/devices/miner-configurator.js:22]: added sending error message to client

**Review by lsheva — APPROVED** [review]

**Pull request merged**